File: python-server/app/core/error_logger.py
"""Async error logger that batches error records into the database."""

import asyncio
import re
from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Optional

from app.core.logging import get_logger

logger = get_logger()

_MAX_BODY_SIZE = 64 * 1024  # 64KB

_SENSITIVE_HEADERS = re.compile(
    r"^(authorization|x-api-key|cookie|set-cookie|proxy-authorization)$",
    re.IGNORECASE,
)


class ErrorCategory:
    PROVIDER_4XX = "provider_4xx"
    PROVIDER_5XX = "provider_5xx"
    TIMEOUT = "timeout"
    NETWORK_ERROR = "network_error"
    CONNECT_ERROR = "connect_error"
    STREAM_ERROR = "stream_error"
    INTERNAL_ERROR = "internal_error"


@dataclass
class ErrorLogRecord:
    error_category: str
    error_message: Optional[str] = None
    error_code: Optional[int] = None
    request_id: Optional[str] = None
    provider_name: Optional[str] = None
    credential_name: Optional[str] = None
    model_requested: Optional[str] = None
    model_mapped: Optional[str] = None
    endpoint: Optional[str] = None
    client_protocol: Optional[str] = None
    provider_protocol: Optional[str] = None
    is_streaming: Optional[bool] = None
    request_body: Optional[str] = None
    response_body: Optional[str] = None
    total_duration_ms: Optional[int] = None
    timestamp: datetime = field(default_factory=lambda: datetime.now(timezone.utc))


def mask_headers(headers: dict) -> dict:
    masked = {}
    for k, v in headers.items():
        if _SENSITIVE_HEADERS.match(k):
            masked[k] = "***"
        else:
            masked[k] = v
    return masked


def _truncate_body(body: Optional[str]) -> Optional[str]:
    if body is None:
        return None
    if len(body) <= _MAX_BODY_SIZE:
        return body
    return body[:_MAX_BODY_SIZE] + "...[truncated]"


class ErrorLogger:
    def __init__(self):
        self._queue: asyncio.Queue[ErrorLogRecord] = asyncio.Queue(maxsize=500)
        self._task: Optional[asyncio.Task] = None
        self._running = False
        self._dropped = 0

    async def start(self):
        self._running = True
        self._task = asyncio.create_task(self._writer_loop())
        logger.info("Error logger started")

    def log_error(self, record: ErrorLogRecord):
        record.request_body = _truncate_body(record.request_body)
        record.response_body = _truncate_body(record.response_body)
        try:
            self._queue.put_nowait(record)
        except asyncio.QueueFull:
            self._dropped += 1
            if self._dropped % 100 == 1:
                logger.warning(f"Error log queue full, dropped {self._dropped} total")

    async def _writer_loop(self):
        buffer: list[ErrorLogRecord] = []
        while self._running or not self._queue.empty():
            try:
                try:
                    record = await asyncio.wait_for(self._queue.get(), timeout=2.0)
                    buffer.append(record)
                    if len(buffer) >= 50:
                        await self._flush(buffer)
                        buffer = []
                except asyncio.TimeoutError:
                    if buffer:
                        await self._flush(buffer)
                        buffer = []
            except asyncio.CancelledError:
                if buffer:
                    await self._flush(buffer)
                break
            except Exception as e:
                logger.error(f"Error in error logger writer loop: {e}")

        if buffer:
            await self._flush(buffer)
        logger.info("Error logger writer task stopped")

    async def _flush(self, buffer: list[ErrorLogRecord]):
        if not buffer:
            return
        try:
            from app.core.database import get_database, ErrorLogModel

            db = get_database()
            if db is None:
                logger.warning("Database not available, discarding error log records")
                return

            async with db.session() as session:
                for rec in buffer:
                    model = ErrorLogModel(
                        timestamp=rec.timestamp,
                        request_id=rec.request_id,
                        error_category=rec.error_category,
                        error_code=rec.error_code,
                        error_message=rec.error_message,
                        provider_name=rec.provider_name,
                        credential_name=rec.credential_name,
                        model_requested=rec.model_requested,
                        model_mapped=rec.model_mapped,
                        endpoint=rec.endpoint,
                        client_protocol=rec.client_protocol,
                        provider_protocol=rec.provider_protocol,
                        is_streaming=rec.is_streaming,
                        request_body=rec.request_body,
                        response_body=rec.response_body,
                        total_duration_ms=rec.total_duration_ms,
                    )
                    session.add(model)
        except Exception as e:
            logger.error(f"Failed to flush error log records: {e}")

    async def shutdown(self):
        self._running = False
        if self._task:
            self._task.cancel()
            try:
                await self._task
            except asyncio.CancelledError:
                pass


_error_logger: Optional[ErrorLogger] = None


async def init_error_logger():
    global _error_logger
    _error_logger = ErrorLogger()
    await _error_logger.start()


def get_error_logger() -> Optional[ErrorLogger]:
    return _error_logger


async def shutdown_error_logger():
    global _error_logger
    if _error_logger:
        await _error_logger.shutdown()
        _error_logger = None


def log_error(
    *,
    error_category: str,
    error_message: Optional[str] = None,
    error_code: Optional[int] = None,
    request_id: Optional[str] = None,
    provider_name: Optional[str] = None,
    credential_name: Optional[str] = None,
    model_requested: Optional[str] = None,
    model_mapped: Optional[str] = None,
    endpoint: Optional[str] = None,
    client_protocol: Optional[str] = None,
    provider_protocol: Optional[str] = None,
    is_streaming: Optional[bool] = None,
    request_body: Optional[str] = None,
    response_body: Optional[str] = None,
    total_duration_ms: Optional[int] = None,
):
    el = get_error_logger()
    if el is None:
        return
    record = ErrorLogRecord(
        error_category=error_category,
        error_message=error_message,
        error_code=error_code,
        request_id=request_id,
        provider_name=provider_name,
        credential_name=credential_name,
        model_requested=model_requested,
        model_mapped=model_mapped,
        endpoint=endpoint,
        client_protocol=client_protocol,
        provider_protocol=provider_protocol,
        is_streaming=is_streaming,
        request_body=request_body,
        response_body=response_body,
        total_duration_ms=total_duration_ms,
    )
    el.log_error(record)

File: rust-server/src/core/error_logger.rs
use axum::http::HeaderMap;
use serde_json::Value;
use sqlx::PgPool;
use std::fmt;
use std::sync::OnceLock;
use tokio::sync::mpsc;

const SENSITIVE_HEADERS: &[&str] = &[
    "authorization",
    "x-api-key",
    "cookie",
    "set-cookie",
    "proxy-authorization",
];

const MAX_BODY_SIZE: usize = 64 * 1024;

pub enum ErrorCategory {
    Provider4xx,
    Provider5xx,
    Timeout,
    NetworkError,
    ConnectError,
    StreamError,
    InternalError,
}

impl fmt::Display for ErrorCategory {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            ErrorCategory::Provider4xx => write!(f, "provider_4xx"),
            ErrorCategory::Provider5xx => write!(f, "provider_5xx"),
            ErrorCategory::Timeout => write!(f, "timeout"),
            ErrorCategory::NetworkError => write!(f, "network_error"),
            ErrorCategory::ConnectError => write!(f, "connect_error"),
            ErrorCategory::StreamError => write!(f, "stream_error"),
            ErrorCategory::InternalError => write!(f, "internal_error"),
        }
    }
}

pub struct ErrorLogRecord {
    pub request_id: String,
    pub error_category: ErrorCategory,
    pub error_message: String,
    pub error_code: Option<i32>,
    pub endpoint: String,
    pub client_protocol: String,
    pub request_headers: Option<Value>,
    pub request_body: Option<Value>,
    pub provider_name: String,
    pub provider_api_base: String,
    pub provider_protocol: String,
    pub mapped_model: String,
    pub response_status_code: Option<i32>,
    pub response_body: Option<Value>,
    pub total_duration_ms: Option<i32>,
    pub credential_name: String,
    pub client: String,
    pub is_streaming: bool,
}

impl Default for ErrorLogRecord {
    fn default() -> Self {
        Self {
            request_id: String::new(),
            error_category: ErrorCategory::InternalError,
            error_message: String::new(),
            error_code: None,
            endpoint: String::new(),
            client_protocol: String::new(),
            request_headers: None,
            request_body: None,
            provider_name: String::new(),
            provider_api_base: String::new(),
            provider_protocol: String::new(),
            mapped_model: String::new(),
            response_status_code: None,
            response_body: None,
            total_duration_ms: None,
            credential_name: String::new(),
            client: String::new(),
            is_streaming: false,
        }
    }
}

pub struct ErrorLogger {
    tx: mpsc::Sender<ErrorLogRecord>,
}

impl ErrorLogger {
    pub fn new(pool: PgPool) -> Self {
        let (tx, rx) = mpsc::channel(500);
        tokio::spawn(Self::writer_task(rx, pool));
        Self { tx }
    }

    pub fn log(&self, record: ErrorLogRecord) {
        if let Err(e) = self.tx.try_send(record) {
            tracing::warn!("Error logger channel full, dropping record: {}", e);
        }
    }

    async fn writer_task(mut rx: mpsc::Receiver<ErrorLogRecord>, pool: PgPool) {
        let mut buffer: Vec<ErrorLogRecord> = Vec::with_capacity(50);
        let mut interval = tokio::time::interval(std::time::Duration::from_secs(2));

        loop {
            tokio::select! {
                Some(record) = rx.recv() => {
                    buffer.push(record);
                    if buffer.len() >= 50 {
                        Self::flush(&pool, &mut buffer).await;
                    }
                }
                _ = interval.tick() => {
                    if !buffer.is_empty() {
                        Self::flush(&pool, &mut buffer).await;
                    }
                }
                else => {
                    if !buffer.is_empty() {
                        Self::flush(&pool, &mut buffer).await;
                    }
                    break;
                }
            }
        }

        tracing::info!("Error logger writer task stopped");
    }

    async fn flush(pool: &PgPool, buffer: &mut Vec<ErrorLogRecord>) {
        if buffer.is_empty() {
            return;
        }

        // Build batch INSERT with numbered parameters
        let count = buffer.len();
        let cols = 18;
        let mut sql = String::from(
            "INSERT INTO error_logs (\
             request_id, error_category, error_message, error_code, \
             endpoint, client_protocol, request_headers, request_body, \
             provider_name, provider_api_base, provider_protocol, mapped_model, \
             response_status_code, response_body, total_duration_ms, \
             credential_name, client, is_streaming\
             ) VALUES ",
        );

        for i in 0..count {
            if i > 0 {
                sql.push_str(", ");
            }
            let base = i * cols + 1;
            sql.push('(');
            for j in 0..cols {
                if j > 0 {
                    sql.push_str(", ");
                }
                sql.push('$');
                sql.push_str(&(base + j).to_string());
            }
            sql.push(')');
        }

        let mut query = sqlx::query(&sql);

        for record in buffer.drain(..) {
            let request_body = truncate_body(&record.request_body);
            let response_body = truncate_body(&record.response_body);

            query = query
                .bind(record.request_id)
                .bind(record.error_category.to_string())
                .bind(record.error_message)
                .bind(record.error_code)
                .bind(record.endpoint)
                .bind(record.client_protocol)
                .bind(record.request_headers)
                .bind(request_body)
                .bind(record.provider_name)
                .bind(record.provider_api_base)
                .bind(record.provider_protocol)
                .bind(record.mapped_model)
                .bind(record.response_status_code)
                .bind(response_body)
                .bind(record.total_duration_ms)
                .bind(record.credential_name)
                .bind(record.client)
                .bind(record.is_streaming);
        }

        if let Err(e) = query.execute(pool).await {
            tracing::error!("Failed to flush error logs to database: {}", e);
        }
    }
}

pub fn mask_headers(headers: &HeaderMap) -> Value {
    let mut map = serde_json::Map::new();
    for (name, value) in headers.iter() {
        let key = name.as_str().to_lowercase();
        let val = if SENSITIVE_HEADERS.contains(&key.as_str()) {
            "***".to_string()
        } else {
            value.to_str().unwrap_or("<non-utf8>").to_string()
        };
        map.insert(key, Value::String(val));
    }
    Value::Object(map)
}

fn truncate_body(body: &Option<Value>) -> Option<Value> {
    let body = body.as_ref()?;
    let serialized = serde_json::to_string(body).ok()?;
    if serialized.len() <= MAX_BODY_SIZE {
        Some(body.clone())
    } else {
        Some(serde_json::json!({
            "_truncated": true,
            "_original_size": serialized.len(),
            "_preview": &serialized[..MAX_BODY_SIZE.min(serialized.len())]
        }))
    }
}

// Global singleton
static ERROR_LOGGER: OnceLock<ErrorLogger> = OnceLock::new();

pub fn init_error_logger(pool: PgPool) {
    let logger = ErrorLogger::new(pool);
    ERROR_LOGGER.set(logger).ok();
    tracing::info!("Error logger initialized");
}

pub fn log_error(record: ErrorLogRecord) {
    if let Some(logger) = ERROR_LOGGER.get() {
        logger.log(record);
    }
}
